### PR TITLE
[SPARK-22848][SQL] Eliminate mutable state from Stack

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR eliminates mutable states from the generated code for `Stack`.

## How was this patch tested?

Existing test suites

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 1]
```diff
@@ -214,11 +214,15 @@ case class Stack(children: Seq[Expression]) extends Generator {
...
-      ev.value,
-      v => s"$v = $wrapperClass$$.MODULE$$.make($rowData);", useFreshName = false)
-    ev.copy(code = code, isNull = "false")
+      "stackWrappedArray",
+      v => s"$v = $wrapperClass$$.MODULE$$.make($rowData);")
+    ev.copy(code =
+      s"""
+         |$code
+         |$wrapperClass<InternalRow> ${ev.value} = $wrappedArray;
+       """.stripMargin, isNull = "false")
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: This change does not use `inline = true` for correct code generation.

**Comment by kiszk** [comment]
This PR comes from [this discussion](https://github.com/apache/spark/pull/19811#discussion_r157408061).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 2]
```diff
@@ -214,11 +213,12 @@ case class Stack(children: Seq[Expression]) extends Generator {
...
     // Create the collection.
     val wrapperClass = classOf[mutable.WrappedArray[_]].getName
-    ctx.addMutableState(
-      s"$wrapperClass<InternalRow>",
-      ev.value,
-      v => s"$v = $wrapperClass$$.MODULE$$.make($rowData);", useFreshName = false)
-    ev.copy(code = code, isNull = "false")
+    ev.copy(code =
+      s"""
+         |InternalRow[] $rowData = new InternalRow[$numRows];
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: this creates a large array every time, and I don't think we have data copy issues for generator expressions...

Comment (kiszk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: I see. I did not imagine that `numRows` is large.
I will revert the code for `rowData`.

**Comment by kiszk** [comment]
Jenkins, retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by viirya** [comment]
LGTM

**Comment by kiszk** [comment]
Jenkins, retest this please

**Comment by viirya** [comment]
I think the test failure is not related to this change, but the ongoing work to upgrade pyarrow.

**Comment by cloud-fan** [comment]
yea it's failing globally, I'm merging this PR, thanks!

**Comment by viirya** [comment]
I ran the test locally. It should be fine as it passes.

**Pull request closed without merging**